segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainerMetadata.java

```diff
@@ -289,15 +293,26 @@ public void setContainerEpoch(long value) {
     @Override
     public Collection<SegmentMetadata> cleanup(Collection<SegmentMetadata> evictionCandidates, long sequenceNumberCutoff) {
         long adjustedCutoff = Math.min(sequenceNumberCutoff, this.lastTruncatedSequenceNumber.get());
+        HashMap<Long, Integer> txnCounts = new HashMap<>();
         Collection<SegmentMetadata> evictedSegments = new ArrayList<>(evictionCandidates.size());
         synchronized (this.lock) {
-            // Find those segments that have active transactions associated with them.
-            Set<Long> activeTransactions = getSegmentsWithActiveTransactions(adjustedCutoff);
+            // Count how many total transactions per parent segment there are.
+            this.metadataById
+                    .values().stream()
+                    .filter(SegmentMetadata::isTransaction)
+                    .forEach(m -> updateValue(txnCounts, m.getParentId(), 1));
+
+            // Subtract the counts for those transactions that will be evicted.
+            evictionCandidates
+                    .stream()
+                    .filter(m -> m.isTransaction() && isEligibleForEviction(m, adjustedCutoff))
+                    .forEach(m -> updateValue(txnCounts, m.getParentId(), -1));
 
             // Remove those candidates that are still eligible for removal and which do not have any active transactions.
             evictionCandidates
                     .stream()
-                    .filter(m -> isEligibleForEviction(m, adjustedCutoff) && !activeTransactions.contains(m.getId()))
+                    .filter(m -> isEligibleForEviction(m, adjustedCutoff)
+                            && txnCounts.getOrDefault(m.getId(), 0) <= 0)
```

Comment:
Makes sense. In the worst case this will only reduce the number of segments evicted, not increase it, so it shouldn't cause issues.